File: 프로그래머스/level3/징검다리 건너기.js
function solution(stones, k) {
    var answer = 0;
    let left = 0;
    let right = 0;
    stones.map(x=>right = Math.max(right, x));
    while(left<=right){
        let mid = Math.floor((right+left)/2);
        let tmp = stones.map(x=>x-mid);
        let flag = 1;
        for(let i=0; i<stones.length; i++){
            if(tmp[i]<=0){
                let flag1 = 0;
                for(let j=i+1; j<=i+k-1; j++){
                    if(tmp[j]>0||j>=tmp.length){
                        flag1 = 1;
                        i=j;
                        break;
                    }
                }
                if(flag1===0){
                    flag = 0;
                    break;
                }
            }
        }
        if(flag === 1){
            left = mid+1;
            answer = mid;
        }
        else{
            right = mid-1;
        }
    }    
    return answer+1;
}

File: 프로그래머스/level3/징검다리 건너기.cpp
#include <string>
#include <algorithm>
#include <vector>
#include <iostream>

using namespace std;

int solution(vector<int> stones, int k) {
    int answer = 0;
    int l = 1;
    int r = 0;
    for(int i=0; i<stones.size(); i++){
        r = max(r, stones[i]);
    }
    while(l<=r){
        int cnt = 0;
        int mid = (l+r)/2;
        int flag = 0;
        for(int i=0; i<stones.size(); i++){
            if(stones[i]-mid+1<=0){
                cnt++;
            }
            else    cnt = 0;
            if(cnt>=k){
                flag = 1;
                break;                
            }
        }
        if(flag == 1){
            r = mid-1;
        }
        else{
            l = mid+1;
            answer = mid;
        }
    }
    return answer;
}